### PR TITLE
test(io): avoid real Azure IO in credential tests

### LINKED ISSUE #1882: test: TestCreateAzureBucketManagedIdentityCredentialCalled is doing real IO
### Apache Iceberg version

None

### Please describe the bug 🐞

and doing a retry loop inside azure sdk (~100s). [An example run](https://github.com/apache/iceberg-go/actions/runs/32902971278/job/97980710396?pr=1679#step:7:7591). 

It's an integration test disguised as unit. It should instead inject credentials factory or HTTP transport. At worst, a shorter context deadline.

cc @alessandro-nori

### PR DESCRIPTION
**Pull request by Revanth14**
## Summary

The Azure credential tests perform real storage operations against a fake account, triggering authentication attempts and SDK retries that can take around 100 seconds.

This change injects credential factories into bucket creation so the tests can verify that the default or managed identity credential path is selected without making network calls. Production credential creation remains unchanged.

Fixes #1882.

## Validation

- `go test ./io/gocloud/... -count=1 -timeout=120s` — passed.

### CODE REVIEW

File: io/gocloud/azure/azure_test.go [hunk 1]
```diff
@@ -23,39 +23,61 @@ import (
...
 	"gocloud.dev/blob"
 	"gocloud.dev/blob/memblob"
 )
 
+func TestProductionAzureCredentialFactories(t *testing.T) {
+	factories := productionAzureCredentialFactories()
+
+	t.Run("default credential", func(t *testing.T) {
+		// Avoid inheriting credential selection or loading credentials from the environment.
+		t.Setenv("AZURE_TOKEN_CREDENTIALS", "dev")
```
Comment (laskoviymishka) [inline io/gocloud/azure/azure_test.go:47]: I'd drop this test, or narrow it right down.

`AZURE_TOKEN_CREDENTIALS` isn't something the Go `azidentity` SDK reads (it's a Java/.NET knob), so this `Setenv` is a no-op and the comment above it is misleading. And `NewDefaultAzureCredential` always returns `*DefaultAzureCredential` on success, so the `IsType` assertion is guaranteed by the signature; it can't fail. The routing is already proven by the two `...CredentialCalled` tests below, so this one only buys the appearance of coverage.

If you want to keep a wiring check, I'd assert the factory fields are non-nil and skip constructing real credentials. wdyt?

File: io/gocloud/azure/azure_test.go [hunk 2]
```diff
@@ -23,39 +23,61 @@ import (
...
-		assert.Contains(t, err.Error(), "DefaultAzureCredential",
-			"Expected DefaultAzureCredential error but got: %v", err)
-		t.Logf("DefaultAzureCredential path was taken and failed at creation: %v", err)
+			return &fake.TokenCredential{}, nil
+		},
+		newManagedIdentity: func(*azidentity.ManagedIdentityCredentialOptions) (azcore.TokenCredential, error) {
+			t.Fatal("managed identity credential factory should not be called")
 
-		return
+			return nil, nil
```
Comment (laskoviymishka) [inline io/gocloud/azure/azure_test.go]: The `t.Fatal` covers us today since both closures run on the test goroutine, but `return nil, nil` is a fragile sentinel: a nil credential with a nil error. If a future refactor ever calls these from a background goroutine, `t.Fatal`'s `Goexit` only kills that goroutine, the return actually executes, and the caller gets a nil credential it thinks is valid.

I'd return an error here instead:

```go
t.Errorf("managed identity credential factory should not be called")
return nil, fmt.Errorf("unexpected call to managed identity credential factory")
```

Same pattern applies to the default-credential closure over in the managed-identity test (line 111).

File: io/gocloud/azure/azure_test.go [hunk 3]
```diff
@@ -81,20 +103,28 @@ func TestCreateAzureBucketManagedIdentityCredentialCalled(t *testing.T) {
...
+	managedIdentityCalled := false
+	credentialFactories := azureCredentialFactories{
+		newDefaultCredential: func(*azidentity.DefaultAzureCredentialOptions) (azcore.TokenCredential, error) {
+			t.Fatal("default credential factory should not be called")
+
+			return nil, nil
+		},
+		newManagedIdentity: func(opts *azidentity.ManagedIdentityCredentialOptions) (azcore.TokenCredential, error) {
+			managedIdentityCalled = true
+			require.Nil(t, opts)
```
Comment (laskoviymishka) [inline io/gocloud/azure/azure_test.go]: `require.Nil(t, opts)` reaches into the production function's local state. If someone ever defaults `miOpts` to `&ManagedIdentityCredentialOptions{}` it'd break this test for no behavioral reason, and `managedIdentityCalled = true` already proves the path was taken.

While we're here: the client-ID branch (`ManagedIdentityCredentialOptions{ID: ...}`) is never exercised. If you added a second sub-case that passes `icebergio.ADLSClientID` and asserted `opts.ID == azidentity.ClientID("...")`, you'd cover that branch and turn this assertion into something meaningful instead of dropping it. wdyt?

**Review by laskoviymishka — CHANGES_REQUESTED** [review]
Nice work on this. The credential-factory injection seam is exactly the right way to kill the IMDS/auth hang, and the two `...CredentialCalled` tests are a real step up from asserting on error strings. They actually prove which branch runs now.

Almost there. One thing I'd fix before merge: `TestProductionAzureCredentialFactories`. The `AZURE_TOKEN_CREDENTIALS=dev` line is a no-op in the Go SDK (that's a Java/.NET knob, not something `azidentity` reads), the comment above it is misleading, and `NewDefaultAzureCredential` always returns `*DefaultAzureCredential` on success, so the `IsType` assertion is guaranteed by the signature and the test doesn't verify anything the routing tests below don't already cover. I'd drop it, or narrow it to a non-nil factory-wiring check that doesn't construct real credentials.

A couple of smaller things I've left inline:

- the `return nil, nil` in the should-not-be-called closures is a fragile sentinel (nil credential, nil error); I'd return an error there instead so a future async caller can't silently continue
- `require.Nil(opts)` couples the test to internal plumbing; it'd be more valuable as a client-ID sub-case that actually exercises the `ManagedIdentityCredentialOptions{ID: ...}` branch

Unrelated to this PR, but while you're in here: `ADLSManagedIdentityEnabled` is a Go-only property with no Java/PyIceberg equivalent, worth a godoc note someday, but not something I'd hold this on.

Fix the one test and this is good to land. Happy to take another pass.

File: io/gocloud/azure/azure_test.go (hunk 1 above)
Comment (Revanth14) [reply io/gocloud/azure/azure_test.go:47]: Thanks for reviewing! I checked our pinned `azidentity v1.14.1`, and it [does read `AZURE_TOKEN_CREDENTIALS`](https://github.com/Azure/azure-sdk-for-go/blob/sdk/azidentity/v1.14.1/sdk/azidentity/default_azure_credential.go#L98-L119). Setting it to `dev` excludes environment, workload identity, and managed identity credentials, and overrides any inherited value. I’ve expanded the comment to clarify that.

The factory returns `azcore.TokenCredential`, so `IsType` checks that our production wrapper constructs the intended credential type. The routing tests inject replacements and wouldn’t catch incorrect production wiring; non-nil checks wouldn’t catch that either.

I’d prefer to keep this small wiring check. It only constructs credentials,there’s no token acquisition or storage IO.

File: io/gocloud/azure/azure_test.go (hunk 2 above)
Comment (Revanth14) [reply io/gocloud/azure/azure_test.go]: Updated both closures to use `t.Errorf` and return a non-nil error via `errors.New`.

One small clarification: `Goexit` terminates the calling goroutine, so the return after `t.Fatal` wouldn’t execute. Still, this makes the failure handling explicit and avoids relying on a fatal assertion inside the callback.

File: io/gocloud/azure/azure_test.go (hunk 3 above)
Comment (Revanth14) [reply io/gocloud/azure/azure_test.go]: Agreed, updated this to cover both cases. The no-client-ID case now accepts nil or empty options and checks that no ID is forwarded. The new client-ID case passes `icebergio.ADLSClientID` and verifies `opts.ID` matches `azidentity.ClientID(clientID)`.

Azure package tests pass.

**Review by laskoviymishka — APPROVED** [review]
LGTM! 👍
